### PR TITLE
Next.js i18n support

### PR DESCRIPTION
**Pull request by leoortizz**
### Description

- [X] Subpath routing support using Firebase Hosting i18n rewrites
- [X] Domain routing support allowing domains registered on Hosting
- [ ] Handle [Next.js NEXT_LOCALE cookie](https://nextjs.org/docs/advanced-features/i18n-routing#leveraging-the-next_locale-cookie)

### CODE REVIEW

File: src/frameworks/utils.ts [hunk 1]
```diff
@@ -39,3 +40,28 @@ export async function warnIfCustomBuildScript(
...
 }
+
+/**
+ * Join the default domain and the custom domains of a Hosting site
+ *
+ * @param projectId the project id
+ * @param siteId the site id
+ * @return array of domains
+ */
+export async function getAllSiteDomains(projectId: string, siteId: string): Promise<string[]> {
```
Comment (leoortizz) [inline src/frameworks/utils.ts:51]: will work on unit tests for this soon

**Review by jamesdaniels — CHANGES_REQUESTED** [review]
Whoops, I did not submit my review from last week 🤦

File: src/frameworks/next/index.ts [hunk 2]
```diff
@@ -229,7 +260,14 @@ export async function build(dir: string): Promise<BuildResult> {
...
   }
 
-  return { wantsBackend, headers, redirects, rewrites, trailingSlash };
+  return {
+    wantsBackend,
+    headers,
+    redirects,
+    rewrites,
+    trailingSlash,
+    ...(!nextjsI18n?.domains ? { i18n: nextjsI18n } : {}),
```
Comment (jamesdaniels) [inline src/frameworks/next/index.ts:269]: Change to just `i18n,` and above let's set `const i18n = nextjsI18n && !nextjsI18n.domains ? nextjsI18n : undefined;` that way we don't need a ternary in a splat

**Review by jamesdaniels — CHANGES_REQUESTED** [review]

**Review by jamesdaniels — APPROVED** [review]
Looks good, I'll give this a spin locally and do some spot testing. Add changelog and I'll merge

File: src/frameworks/next/index.ts [hunk 3]
```diff
@@ -234,7 +256,16 @@ export async function build(dir: string): Promise<BuildResult> {
     console.log("");
   }
 
-  return { wantsBackend, headers, redirects, rewrites, trailingSlash };
+  const i18n = nextjsI18n && !nextjsI18n.domains ? nextjsI18n : undefined;
```
Comment (TheIronDev) [inline src/frameworks/next/index.ts:259]: Nit: Maybe add parens to make this easier to read?

`  const i18n = (nextjsI18n && !nextjsI18n.domains) ? nextjsI18n : undefined;`

**Review by TheIronDev — APPROVED** [review]

File: src/frameworks/next/index.ts (hunk 3 above)
Comment (leoortizz) [reply src/frameworks/next/index.ts:259]: @TheIronDev prettier does not allow parenthesis there. Maybe a variable name could help? e.g:
```ts
  const isUsingI18nButNotDomainRouting = nextjsI18n && !nextjsI18n.domains;
  const i18n = isUsingI18nButNotDomainRouting ? nextjsI18n : undefined;
```

**Comment by jamesdaniels** [comment]
Closing in favor of #5788

**Pull request closed without merging**